external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -808,3 +920,124 @@ fn exp(context: &mut Context, e: &T::Exp) {
+    if !is_defined_in_current_module {
+        let msg = format!(
+            "Invalid event. The function '{}::{}' must be called with a type defined in the current module, '{}'",
+            module, name, current_module
+        );
+        let ty_msg = format!(
+            "The type {} is not declared in the current module, '{}'",
+            error_format(first_ty, &Subst::empty()),
+            current_module
+        );
+        context.env.add_diag(diag!(
+            EVENT_EMIT_CALL_DIAG,
+            (loc, msg),
+            (first_ty.loc, ty_msg)
+        ));
+    }
+}
+
+fn check_private_transfer(context: &mut Context, loc: Loc, mcall: &ModuleCall) {
+    let ModuleCall {
+        module,
+        name,
+        type_arguments,
+        ..
+    } = mcall;
+    let current_module = context.current_module();
+    if current_module
+        .value
+        .is(SUI_ADDR_NAME, TRANSFER_FUNCTION_NAME)
+    {
+        // inside the transfer module, so no private transfer rules
+        return;
+    }
+    let Some(first_ty) = type_arguments.first() else {
+        // invalid arity
+        debug_assert!(false, "ICE arity should have been expanded for errors");
+        return;
+    };
+    let (in_current_module, first_ty_tn) = match first_ty.value.type_name() {
+        Some(sp!(_, TypeName_::Multiple(_))) | Some(sp!(_, TypeName_::Builtin(_))) | None => {
+            (false, None)
+        }
+        Some(sp!(_, TypeName_::ModuleType(m, n))) => (m == current_module, Some((m, n))),
+    };
+    if !in_current_module {
+        let mut msg = format!(
+            "Invalid private transfer. \
+            The function '{}::{}' is restricted to being called in the object's module",
+            module, name,
+        );
+        if let Some((first_ty_module, _)) = &first_ty_tn {
+            msg = format!("{}, '{}'", msg, first_ty_module);
+        };
+        let ty_msg = format!(
+            "The type {} is not declared in the current module, '{}'",
+            error_format(first_ty, &Subst::empty()),
+            current_module
+        );
+        let mut diag = diag!(
+            PRIVATE_TRANSFER_CALL_DIAG,
+            (loc, msg),
+            (first_ty.loc, ty_msg)
+        );
+        if first_ty
+            .value
+            .has_ability_(Ability_::Store)
+            .is_some_and(|b| b)
+        {
+            let store_loc = if let Some((first_ty_module, first_ty_name)) = &first_ty_tn {
+                let abilities = context
+                    .info
+                    .struct_declared_abilities(first_ty_module, first_ty_name);
+                abilities.ability_loc_(Ability_::Store).unwrap()
+            } else {
+                first_ty
+                    .value
+                    .abilities(first_ty.loc)
+                    .unwrap()
+                    .ability_loc_(Ability_::Store)
```

Comment:
```suggestion
                    .and_then(|a| a.ability_loc_(Ability_::Store))
```

Reply:
I would like the compiler to crash if this ever changes. I mean really there should be an `ICE` or something there, that is a bit more work than I think it is worth for this case, so I'll at least change it to an `expect`